numpy/_core/src/multiarray/arrayobject.c

```diff
@@ -358,56 +358,61 @@ PyArray_ResolveWritebackIfCopy(PyArrayObject * self)
 /*********************** end C-API functions **********************/
 
 
-/* dealloc must not raise an error, best effort try to write
-   to stderr and clear the error
+/* if unraisable, called from dealloc, which must not raise an error,
+   so make best effort try to write to stderr and clear the error.
 */
 
-static inline void
-WARN_IN_DEALLOC(PyObject* warning, const char * msg) {
-    if (PyErr_WarnEx(warning, msg, 1) < 0) {
-        PyObject * s;
-
-        s = PyUnicode_FromString("array_dealloc");
-        if (s) {
-            PyErr_WriteUnraisable(s);
-            Py_DECREF(s);
-        }
-        else {
-            PyErr_WriteUnraisable(Py_None);
-        }
+static inline int
+WARN_IN_DEALLOC(PyObject* warning, const char * msg, npy_bool unraisable) {
+    int ret = PyErr_WarnEx(warning, msg, 1);
+    if (ret == 0 || !unraisable) {
+        return ret;
     }
+    PyObject *s = PyUnicode_FromString("array_dealloc");
+    PyErr_WriteUnraisable(s);  // OK if s = NULL
+    Py_XDECREF(s);
+    return 0;
 }
 
 /* array object functions */
 
-static void
-array_dealloc(PyArrayObject *self)
+/*
+ * Much of the actual work for dealloc, split off for use in __setstate__
+ * via dealloc_all_but_self function defined below.
+ * If not unraisable, will return -1 on error, 0 on success.
+ * If unraisable, always succeeds.
+ */
+static int
+_dealloc_all_but_self(PyArrayObject *self, npy_bool unraisable)
 {
     PyArrayObject_fields *fa = (PyArrayObject_fields *)self;
+    int retval;
 
-    if (_buffer_info_free(fa->_buffer_info, (PyObject *)self) < 0) {
-        PyErr_WriteUnraisable(NULL);
+    if (fa->_buffer_info) {
+        if (_buffer_info_free(fa->_buffer_info, (PyObject *)self) < 0) {
+            if (!unraisable) {
+                return -1;
+            }
+            PyErr_WriteUnraisable(NULL);
+        }
+        fa->_buffer_info = NULL;
     }
 
-    if (fa->weakreflist != NULL) {
-        PyObject_ClearWeakRefs((PyObject *)self);
-    }
     if (fa->base) {
-        int retval;
         if (PyArray_FLAGS(self) & NPY_ARRAY_WRITEBACKIFCOPY)
         {
             char const * msg = "WRITEBACKIFCOPY detected in array_dealloc. "
                 " Required call to PyArray_ResolveWritebackIfCopy or "
                 "PyArray_DiscardWritebackIfCopy is missing.";
-            /*
-             * prevent reaching 0 twice and thus recursing into dealloc.
```

Comment:
I don't understand at all why this was there -- if there is a piece of code that tries to defref self again, it is a bug, and should be exposed. Given that tests pass without it, it doesn't seem to be necessary (any more?).

Reply:
It is probably never needed, but please don't remove it. If we remove it, I would rather just make the whole pass a full error, since the warning has been around forever (yes, it's not a raised error unfortunately, as it it will be printed).

The problem this protects against is if `PyArray_ResolveWritebackIfCopy` calls `Py_INCREC(self)` for any reason at all, things would blow up.
And while it may be that there is no path that will do an incref, it seems opaque enough to follow that leaking a global reference here is immaterial.
(Yeah, I am basically saying that I don't want to audit/promise that `PyArray_ResolveWritebackIfCopy` won't cause an `INCREF`.)